### PR TITLE
Add Featurizer for RFDiffusion added

### LINKED ISSUE #4980: Adding Protein Backbone Featurizer for RFDiffusion Model
## 🚀 Feature

I have added a Featurizer that extracts protein backbone coordinates for a residue sequence and converts it into:
1. Sequential Embeddings
2. Pairwise Embeddings and
3. Frames consisting of Rotation+Translation matrices

It is done as a part 1/12 of my GSoC 2026 proposal for integrating RFdiffusion model pipeline into the deepchem library.

## Changes

It adds two files:

1. deepchem/feat/RFDiffusion_featurizer.py
2. deepchem/feat/test_RFDiffusion_featurizer.py

Where the featurizer performs following operations:

- Extracts the coordinates of N, Ca, C backbones of amino acid sequences, giving the result as (N, 3, 3)
- Converts the backbone coordinates into Rotational and Translational frames, later to be used for 3D track diffusion process
- Uses the frames to create 2D Pairwise embeddings (N, N, d_emb), which will be used for 2D track diffusion process
- It also creates a learnable 1D embedding which is used along with a sinusoidal time based embedding

## Tests

I have created a wide range of test cases to test the robustness of featurizer functions to edge cases, general cases, as well as missing value cases.

##Feedback needed

- On quality of test cases and what could be done to improve them
- Method used to generate frames and embeddings
- Quality of code or if it needs any improvement

## References

Watson et al. "De novo design of protein structure and function with RFdiffusion." Nature 2023
Ho et al. "Denoising diffusion probabilistic models." NeurIPS 2020
Nichol and Dhariwal "Improved denoising diffusion probabilistic models." ICML 2021

### PR DESCRIPTION
**Pull request by shubham-bari**
## Description

Fix #4980 

I have added a Featurizer that extracts protein backbone coordinates for a residue sequence and converts it into:

- Sequential Embeddings
- Pairwise Embeddings and
- Frames consisting of Rotation+Translation matrices

It is done as a part 1/12 of my GSoC 2026 proposal for integrating RFdiffusion model pipeline into the deepchem library.

It adds two files:

1. deepchem/feat/RFDiffusion_featurizer.py
2. deepchem/feat/test_RFDiffusion_featurizer.py

Where the featurizer performs following operations:

- Extracts the coordinates of N, Ca, C backbones of amino acid sequences, giving the result as (N, 3, 3)
- Converts the backbone coordinates into Rotational and Translational frames, later to be used for 3D track diffusion process
- Uses the frames to create 2D Pairwise embeddings (N, N, d_emb), which will be used for 2D track diffusion process
- It also creates a learnable 1D embedding which is used along with a sinusoidal time based embedding

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/feat/RFDiffusion_featurizer.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        Returns
+        -------
+        frames: np.ndarray
+            Array of shape ``N, 3, 3`` which represents the frames for the backbone coordinates.
+            N: Number of residues in the protein.
+            3: Number of atoms per residue (N, Ca, C).
+            3: Number of coordinates per atom (x, y, z).
+        """
+        if verbose:
+            print("Computing frames for backbone coordinates...")
```
Comment (Harindhar10) [inline deepchem/feat/RFDiffusion_featurizer.py:335]: Use logger instead of print statements

File: deepchem/utils/pytorch_utils.py [hunk 2]
```diff
@@ -636,3 +636,42 @@ def get_complex_dtype(dtype: torch.dtype) -> torch.dtype:
...
+    -------
+    torch.Tensor
+        Sinusoidal embeddings for values.
+
+    Examples
+    --------
+    >>> import torch
+    >>> from deepchem.utils.pytorch_utils import get_sinusoid
+    >>> values = torch.tensor([1.0, 2.0, 3.0])
+    >>> get_sinusoid(values, 10.0, 32)
```
Comment (Harindhar10) [inline deepchem/utils/pytorch_utils.py]: Add the shape of the output below

File: deepchem/utils/pytorch_utils.py [hunk 3]
```diff
@@ -636,3 +636,42 @@ def get_complex_dtype(dtype: torch.dtype) -> torch.dtype:
     else:
         raise TypeError("Unsupported datatype %s for conversion to complex" %
                         dtype)
+
+
+def get_sinusoid(values, max_val, channels=32):
+    """Get sinusoidal embeddings for values.
+
```
Comment (Harindhar10) [inline deepchem/utils/pytorch_utils.py]: Please add a more descriptive doc string explaining what sinusoidal embeddings are

File: deepchem/utils/pytorch_utils.py [hunk 4]
```diff
@@ -636,3 +636,42 @@ def get_complex_dtype(dtype: torch.dtype) -> torch.dtype:
...
+    >>> import torch
+    >>> from deepchem.utils.pytorch_utils import get_sinusoid
+    >>> values = torch.tensor([1.0, 2.0, 3.0])
+    >>> get_sinusoid(values, 10.0, 32)
+
+    Returns
+    -------
+    torch.Tensor
+        Sinusoidal embeddings for values.
+    """
```
Comment (Harindhar10) [inline deepchem/utils/pytorch_utils.py]: Please add a reference here. Is it the same sinusoidal positional embedding used in the paper "Attention is all you need"?

File: deepchem/utils/pytorch_utils.py [hunk 5]
```diff
@@ -636,3 +636,42 @@ def get_complex_dtype(dtype: torch.dtype) -> torch.dtype:
     else:
         raise TypeError("Unsupported datatype %s for conversion to complex" %
                         dtype)
+
+
```
Comment (Harindhar10) [inline deepchem/utils/pytorch_utils.py]: Is pytorch_utils the appropriate file for this function, or is it better to create another file called attention_utils or rf_diffusion_utils?

File: docs/source/api_reference/featurizers.rst [hunk 6]
```diff
@@ -369,7 +369,7 @@ SAMFeaturizer
 
 .. autoclass:: deepchem.feat.SAMFeaturizer
   :members:
-  :inherited-members: 
+  :inherited-members:
```
Comment (Harindhar10) [inline docs/source/api_reference/featurizers.rst:372]: Is this deletion of space necessary?

File: deepchem/feat/RFDiffusion_featurizer.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+          shape: [batch_size, R, t]
+          R: Rotation matrix of shape [batch_size, num_residue, 3, 3]
+          t: Translation matrix of shape [batch_size, num_residue, 3]
+
+        4. The frames computed from the N, Ca, C backbone atom coordinates, are used
+           to compute the features for the 2D track of the RFDiffusion model, which
+           capture the pairwise distances and orientations between the backbone atoms.
+
+        shape: [batch_size, num_residues, num_residues, d_pair]
```
Comment (Harindhar10) [inline deepchem/feat/RFDiffusion_featurizer.py:51]: Please add usage example and reference

File: docs/source/api_reference/featurizers.rst (hunk 6 above)
Comment (shubham-bari) [reply docs/source/api_reference/featurizers.rst:372]: There was some error with flake8 tests thats why i had to rewrite it (worked properly after)

File: deepchem/utils/pytorch_utils.py (hunk 4 above)
Comment (shubham-bari) [reply deepchem/utils/pytorch_utils.py]: Yes. I will add a reference to this

File: deepchem/utils/pytorch_utils.py (hunk 2 above)
Comment (shubham-bari) [reply deepchem/utils/pytorch_utils.py]: done

File: deepchem/feat/RFDiffusion_featurizer.py (hunk 7 above)
Comment (shubham-bari) [reply deepchem/feat/RFDiffusion_featurizer.py:51]: Added shape along with usage example

File: deepchem/feat/RFDiffusion_featurizer.py (hunk 1 above)
Comment (shubham-bari) [reply deepchem/feat/RFDiffusion_featurizer.py:335]: Done

File: deepchem/utils/pytorch_utils.py (hunk 3 above)
Comment (shubham-bari) [reply deepchem/utils/pytorch_utils.py]: Added a descriptive docstring along with references
.

File: deepchem/utils/pytorch_utils.py (hunk 4 above)
Comment (shubham-bari) [reply deepchem/utils/pytorch_utils.py]: Reference added along with a more desceriptive docstring

File: deepchem/utils/pytorch_utils.py (hunk 5 above)
Comment (shubham-bari) [reply deepchem/utils/pytorch_utils.py]: I was confused regarding that, and yes pytorch utils are not the correct ones, I have made a separate utils file for our function

**Review by shubham-bari — COMMENTED** [review]
Necessary changes made
